<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Simulate models &mdash; cameo 0.6.0+1.g10719db.dirty documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.6.0+1.g10719db.dirty',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="_static/favicon.ico"/>
    <link rel="top" title="cameo 0.6.0+1.g10719db.dirty documentation" href="index.html" />
    <link rel="up" title="Tutorials" href="tutorials.html" />
    <link rel="next" title="Analyzing models" href="04-analyze-models.html" />
    <link rel="prev" title="Import models" href="02-import-models.html" />
   
  
    <link rel="apple-touch-icon" href="_static/cameo_touch_icon.png" />
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="simulate-models">
<h1>Simulate models<a class="headerlink" href="#simulate-models" title="Permalink to this headline">¶</a></h1>
<p>cameo uses and extends the model data structures defined by
<a class="reference external" href="https://opencobra.github.io/cobrapy/">cobrapy</a>, our favorite
COnstraints-Based Reconstruction and Analysis tool for Python. cameo is
thus 100% compatible with cobrapy. For efficiency reasons, however,
cameo implements its own simulation methods that take advantage of a
more advanced solver interface.</p>
<div class="section" id="primer-constraint-based-modeling">
<h2>Primer: Constraint-Based Modeling<a class="headerlink" href="#primer-constraint-based-modeling" title="Permalink to this headline">¶</a></h2>
<p>Constraint-based modeling is a powerful modeling framework for analyzing
metabolism on the genome scale (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23632383">McCloskey et al.,
2013</a>). For a model that
encompasses <span class="math">\(n\)</span> reactions that involve <span class="math">\(m\)</span> metabolites,
<span class="math">\(\mathbf{S}\)</span> is a matrix of dimension <span class="math">\(m \times n\)</span> that
encodes the stoichiometry of the metabolic reaction system; it is
usually referred to as stoichiometric matrix. Assuming that the system
is in a steady state—the concentration of metabolites are constant—the
system of flux-balances can be formulated as</p>
<div class="math">
\[\begin{align}
\mathbf{S} \mathbf{v} = 0\,,
\end{align}\]</div>
<p>where <span class="math">\(\mathbf{v}\)</span> is the vector of flux rates. With the addition
of a biologically meaningful objective, flux capacity constraints,
information about the reversibility of reactions under physiological
conditions, an optimization problem can be formulated that can easily be
solved using <a class="reference external" href="https://en.wikipedia.org/wiki/Linear_programming">linear
programming</a>.</p>
<p>, e.g., maximimization of biomass production,Given the maximization of
growth rate as one potential biological objective <span class="math">\(v_{biomass}\)</span>,
i.e., the flux of an artificial reaction that consumes biomass
components in empirically determined proportions, and assuming that the
cell is evolutionary optimized to achieve that objective, and
incorporating knowledge about reaction reversibility, uptake and
secretion rates, and maximum flux capacities in the form of lower and
uppers bounds (<span class="math">\(\mathbf{v}_{lb}\)</span> and <span class="math">\(\mathbf{v}_{ub}\)</span>) on
the flux variables <span class="math">\(\mathbf{v}\)</span>, one can formulate and solve an
optimization problem to identify an optimal set of flux rates using flux
balance analysis (FBA):</p>
<div class="math">
\[\begin{split}\begin{align}
 Max ~ &amp; ~ Z_{obj} = \mathbf{c}^{T} \mathbf{v}\\
 \text{s.t.}~ &amp; ~ \mathbf{S} \mathbf{v} = 0 \\
 ~ &amp; ~ \mathbf{v}_{lb} \leq \mathbf{v} \leq \mathbf{v}_{ub} \,.
\end{align}\end{split}\]</div>
</div>
<div class="section" id="flux-balance-analysis">
<h2>Flux Balance Analysis<a class="headerlink" href="#flux-balance-analysis" title="Permalink to this headline">¶</a></h2>
<p>Load a model.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cameo</span> <span class="kn">import</span> <span class="n">load_model</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">load_model</span><span class="p">(</span><span class="s1">&#39;iJO1366&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In cameo, flux balance analysis can be performed with the function
<code class="xref py py-obj docutils literal"><span class="pre">fba</span></code>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span></span>from cameo import fba
%time fba_result = fba(model)
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span>CPU times: user 410 ms, sys: 6.43 ms, total: 416 ms
Wall time: 421 ms
</pre></div>
</div>
<p>Basically, <code class="xref py py-obj docutils literal"><span class="pre">fba</span></code> calls <code class="xref py py-obj docutils literal"><span class="pre">model.solve()</span></code> and wraps the optimization
solution in a <code class="xref py py-obj docutils literal"><span class="pre">FluxDistributionResult</span></code> object. The maximum objective
values (corresponding to a maximum growth rate) can obtained throug
<code class="xref py py-obj docutils literal"><span class="pre">result.objective_value</span></code>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span></span><span class="n">fba_result</span><span class="o">.</span><span class="n">data_frame</span>
</pre></div>
</div>
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>flux</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>DM_4crsol_c</th>
      <td>0.000219</td>
    </tr>
    <tr>
      <th>DM_5drib_c</th>
      <td>0.000221</td>
    </tr>
    <tr>
      <th>DM_aacald_c</th>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>DM_amob_c</th>
      <td>0.000002</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>ZN2t3pp</th>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>ZN2tpp</th>
      <td>0.000335</td>
    </tr>
    <tr>
      <th>ZNabcpp</th>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Zn2tex</th>
      <td>0.000335</td>
    </tr>
  </tbody>
</table>
<p>2583 rows × 1 columns</p>
</div></div>
<div class="section" id="parsimonious-flux-balance-analysis">
<h2>Parsimonious Flux Balance Analysis<a class="headerlink" href="#parsimonious-flux-balance-analysis" title="Permalink to this headline">¶</a></h2>
<p>Parsimonious flux balance analysis (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/20664636">Lewis et al.,
2010</a>), a variant of
FBA, performs FBA in in a first step to determine the maximum objective
value <span class="math">\(Z_{obj}\)</span>, fixes it in form of an additional model
constraint (<span class="math">\(\mathbf{c}^{T} \mathbf{v} \ge Z_{obj}\)</span>), and then
minimizes in a second optimization the <span class="math">\(L_1\)</span> norm of
<span class="math">\(\mathbf{v}\)</span>. The assumption behind pFBA is that cells try to
minimize flux magnitude as well in order to keep protein costs low.</p>
<div class="math">
\[\begin{split}\begin{align}
 Max ~ &amp; ~ \lvert \mathbf{v} \rvert\\
 \text{s.t.}~ &amp; ~ \mathbf{S} \mathbf{v} = 0 \\
 &amp; ~ \mathbf{c}^{T} \mathbf{v} \ge Z_{obj} \\
 ~ &amp; ~ \mathbf{v}_{lb} \leq \mathbf{v} \leq \mathbf{v}_{ub} \,.
\end{align}\end{split}\]</div>
<p>In cameo, pFBA can be performed with the function <code class="xref py py-obj docutils literal"><span class="pre">pfba</span></code>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span></span>from cameo import pfba
%time pfba_result = pfba(model)
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span>CPU times: user 429 ms, sys: 19 ms, total: 448 ms
Wall time: 460 ms
</pre></div>
</div>
<p>The <code class="xref py py-obj docutils literal"><span class="pre">objective_function</span></code> value is <span class="math">\(\lvert \mathbf{v} \rvert\)</span> ...</p>
<div class="code python highlight-python"><div class="highlight"><pre><span></span><span class="n">pfba_result</span><span class="o">.</span><span class="n">objective_value</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="mf">699.0222751839472</span>
</pre></div>
</div>
<p>... whis is significantly smaller than flux vector of the original FBA
solution.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span></span><span class="nb">abs</span><span class="p">(</span><span class="n">fba_result</span><span class="o">.</span><span class="n">data_frame</span><span class="o">.</span><span class="n">flux</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="mf">702.81946544045991</span>
</pre></div>
</div>
</div>
<div class="section" id="setp-2-simulate-knockouts-phenotypes">
<h2>Setp 2: Simulate knockouts phenotypes<a class="headerlink" href="#setp-2-simulate-knockouts-phenotypes" title="Permalink to this headline">¶</a></h2>
<p>Although PFBA and FBA can be used to simulate the effect of knockouts,
other methods have been proven more valuable for that task: MOMA and
ROOM. In <em>cameo</em> we implement a linear version of MOMA.</p>
<hr class="docutils" />
<p>Simulating knockouts:</p>
<ul class="simple">
<li>Manipulate the bounds of the reaction (or use the shorthand method
knock_out)</li>
</ul>
<div class="code python highlight-python"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">PGI</span>
</pre></div>
</div>
<table>
    <tr>
        <td><strong>Id</strong></td><td>PGI</td>
    </tr>
    <tr>
        <td><strong>Name</strong></td><td>Glucose-6-phosphate isomerase</td>
    </tr>
    <tr>
        <td><strong>Stoichiometry</strong></td><td>g6p_c <=> f6p_c</td>
    </tr>
    <tr>
        <td><strong>Lower bound</strong></td><td>-1000.000000</td>
    </tr>
    <tr>
        <td><strong>Upper bound</strong></td><td>1000.000000</td>
    </tr>
</table><div class="code python highlight-python"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">PGI</span><span class="o">.</span><span class="n">knock_out</span><span class="p">()</span>
<span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="o">.</span><span class="n">PGI</span>
</pre></div>
</div>
<table>
    <tr>
        <td><strong>Id</strong></td><td>PGI</td>
    </tr>
    <tr>
        <td><strong>Name</strong></td><td>Glucose-6-phosphate isomerase</td>
    </tr>
    <tr>
        <td><strong>Stoichiometry</strong></td><td>g6p_c --> f6p_c</td>
    </tr>
    <tr>
        <td><strong>Lower bound</strong></td><td>0.000000</td>
    </tr>
    <tr>
        <td><strong>Upper bound</strong></td><td>0.000000</td>
    </tr>
</table><ul class="simple">
<li>Simulate using different methods:</li>
</ul>
<div class="code python highlight-python"><div class="highlight"><pre><span></span>%time fba_knockout_result = fba(model)
fba_knockout_result[model.reactions.BIOMASS_Ec_iJO1366_core_53p95M]
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span>CPU times: user 266 ms, sys: 4.96 ms, total: 271 ms
Wall time: 272 ms
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="mf">0.9761293262947403</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span></span>%time pfba_knockout_result = pfba(model)
pfba_knockout_result[model.reactions.BIOMASS_Ec_iJO1366_core_53p95M]
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span>CPU times: user 374 ms, sys: 3.5 ms, total: 378 ms
Wall time: 379 ms
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="mf">0.9761293262947374</span>
</pre></div>
</div>
<p>MOMA and ROOM relly on a reference (wild-type) flux distribution and we
can use the one previously computed.</p>
<p><strong>Parsimonious FBA references seem to produce better results using this
methods</strong></p>
<div class="code python highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cameo.flux_analysis.simulation</span> <span class="kn">import</span> <span class="n">room</span><span class="p">,</span> <span class="n">lmoma</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span></span>%time lmoma_result = lmoma(model, reference=pfba_result.fluxes)
lmoma_result[model.reactions.BIOMASS_Ec_iJO1366_core_53p95M]
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span>CPU times: user 16.3 s, sys: 246 ms, total: 16.6 s
Wall time: 16.8 s
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="mf">0.8724093536243601</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span></span>%time room_result = room(model, reference=pfba_result.fluxes)
room_result[model.reactions.BIOMASS_Ec_iJO1366_core_53p95M]
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/cameo_logo.png" alt="Logo"/>
    
  </a>
</p>





<p>
<iframe src="https://ghbtns.com/github-btn.html?user=biosustain&repo=cameo&type=watch&count=true&size=large"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>




    

<p>
<a href="https://travis-ci.org/biosustain/cameo">
    <img
        alt="https://secure.travis-ci.org/biosustain/cameo.png?branch=master"
        src="https://secure.travis-ci.org/biosustain/cameo.png?branch=master"
    />
</a>
</p>




    

<p>
<a href="https://codecov.io/github/biosustain/cameo">
    <img
    alt="https://codecov.io/github/biosustain/cameo/coverage.svg?branch=master"
    src="https://codecov.io/github/biosustain/cameo/coverage.svg?branch=master"
    />
</a>
</p>
<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">FAQ</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="08-high-level-API.html">High-level interface for users</a></li>
<li class="toctree-l2"><a class="reference internal" href="02-import-models.html">Import models</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="">Simulate models</a></li>
<li class="toctree-l2"><a class="reference internal" href="04-analyze-models.html">Analyzing models</a></li>
<li class="toctree-l2"><a class="reference internal" href="05-predict-gene-knockout-strategies.html">Predict gene knockout strategies</a></li>
<li class="toctree-l2"><a class="reference internal" href="06-predict-gene-modulation-targets.html">Predict expression modulation targets</a></li>
<li class="toctree-l2"><a class="reference internal" href="07-predict-heterologous-pathways.html">Predict heterologous pathways</a></li>
<li class="toctree-l2"><a class="reference internal" href="09-vanillin-production.html">Vanillin production</a></li>
<li class="toctree-l2"><a class="reference internal" href="09-vanillin-production.html#genome-scale-metabolic-model">Genome-scale metabolic model</a></li>
<li class="toctree-l2"><a class="reference internal" href="09-vanillin-production.html#heterologous-pathway">Heterologous pathway</a></li>
<li class="toctree-l2"><a class="reference internal" href="12-advanced-usage-of-heuristic-optimization.html">Using meta-heuristics to search for knockout strategies.</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to cameo</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="API.html">API Docs</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="tutorials.html">Tutorials</a><ul>
      <li>Previous: <a href="02-import-models.html" title="previous chapter">Import models</a></li>
      <li>Next: <a href="04-analyze-models.html" title="next chapter">Analyzing models</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2014, Novo Nordisk Foundation Center for Biosustainability.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.5</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
      |
      <a href="_sources/03-simulate-models.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/biosustain/cameo" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>